### DISCUSSION TITLE
Heap, Stack, Boxing and Unboxing, Performance ... let's order things!

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#4] [General]
# Heap, Stack, Boxing and Unboxing, Performance ... let's order things!

In this article I will shade some lights on some of the most used terms which seems very confusing especially for beginners: heap, stack and boxing and unboxing.
Furthermore we will also encounter internet wisdom like:
Value types get stored on the stack. Reference types on the heap
We discuss why this is wrong and what the hell performance has to do with it?

https://steven-giesel.com/blogPost/9a40d278-9a9f-49fe-bbfd-2d813a58e73e

**Comment by sabitkondakci** [comment]
##### Thanks for sharing, it was a nice read, well done!

I would like to add a subtle detail on "stack";

Counterintuitively stack grows downward even though we tend to visualize it - like a stack designed by arrays, last item is of the highest **index** or **memory address** - the truth is that last item put on "stack frame" will be having the lowest memory address.

> Default stack size is 2MB on Windows.

```csharp
public struct Model
{
     int a = 10; // memory address on stack in decimal: 2866988798
     int b = 20; // memory address on stack in decimal: 2866988794 

     // storing the reference address on stack
     ItemClass c = new ItemClass(); // memory address on stack in decimal: 2866988790
}
```

**Reply by linkdotnet** [reply]
Thanks @SabitKondakci for sharing. I liked that kind of information. Even though in preparation I played around the memory profiler in Rider it never come to my mind that the memory addresses are the other way around :)

**Comment by fadyfayek11** [comment]
In C# records is a value type, but when it converts to IL it's been as a class! and classes are reference types.
Can you explain it to me!

**Reply by linkdotnet** [reply]
Hey @fadyfayek11. Records itself are not value types. They are indeed reference types. Described [here](https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/tutorials/records). 

If you want to have value type records you have to use: `record struct`, which were introduced with C# 10.
```csharp
public record struct MyDate(int Ticks);
```
You can check [Sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgPgAQEwAYCwAoZBGTmYRgCmAzgA4QDGRABCQC4AmA3HlkgMyYDemN/NZEj4DeGATQC+maRkwIONAE5FKAeyWM69JQFdK9GgFkAngBEI9IgAoANmpgBzGgBUolANYkAlMyA===) how this gets translated.

More information here: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-10.0/record-structs

Bottom line: `record MyRecord` is translated into a `class` aka reference type where as `record struct MyRecord` will be translated into a `struct` aka value type.
